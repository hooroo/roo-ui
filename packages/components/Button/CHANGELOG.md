# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.0.3"></a>
## [0.0.3](https://github.com/hooroo/roo-ui/compare/v0.0.2...v0.0.3) (2018-02-01)




**Note:** Version bump only for package @roo-ui/button

<a name="0.0.2"></a>
## [0.0.2](https://github.com/hooroo/roo-ui/compare/v0.0.1...v0.0.2) (2018-02-01)




**Note:** Version bump only for package @roo-ui/button

<a name="0.0.1"></a>
## 0.0.1 (2018-02-01)




**Note:** Version bump only for package @roo-ui/button
