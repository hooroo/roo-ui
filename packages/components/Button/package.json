{
  "name": "@roo-ui/button",
  "version": "0.1.0",
  "main": "dist/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "babel src -d dist"
  },
  "peerDependencies": {
    "prop-types": "^15.6.0",
    "react": "^16.2.0",
    "react-dom": "^16.2.0",
    "styled-components": "^3.1.5"
  }
}
