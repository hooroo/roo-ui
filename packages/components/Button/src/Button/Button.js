import system from 'system-components';
import { css } from 'styled-components';
import { themeGet, complexStyle } from 'styled-system';
import { darken } from 'polished';

const background = (props) => {
  const styles = complexStyle({
    prop: 'buttonStyle',
    key: 'buttons',
  })(props);

  return props.bg
    ? themeGet(`colors.${props.bg}`, props.bg)(props)
    : styles.backgroundColor;
};

const hoverBackground = props => darken(0.1, background(props));

const Button = system({
  is: 'button',
  buttonStyle: 'default',
  boxShadow: null,
  color: null,
  bg: null,
  border: 2,
  borderColor: 'transparent',
  borderRadius: 'default',
  fontSize: 'sm',
  fontWeight: 'bold',
  letterSpacing: 'wide',
  lineHeight: 'normal',
  px: 5,
  py: 3,
  m: 0,
});

export default Button.extend`
  font-family: ${themeGet('fontFamily')};
  transition: ${themeGet('transitions.default')};
  display: inline-block;
  cursor: pointer;
  outline: 0;
  appearance: none;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;

  &:focus {
    box-shadow: ${themeGet('shadows.focus')};
  }

  &:hover {
    background-color: ${hoverBackground};
  }

  &[disabled] {
    opacity: 0.7;
    cursor: not-allowed;
  }

  ${props => props.rounded && css`
    border-radius: ${themeGet('radii.rounded')};
  `}

  ${props => props.outlined && css`
    background-color: transparent;
    color: ${background};
    border-color: ${background};

    &:hover {
      background-color: transparent;
      color: ${hoverBackground};
      border-color: ${hoverBackground};
    }
  `}
`;
