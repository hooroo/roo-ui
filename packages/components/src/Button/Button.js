import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import { darken } from 'polished';
import tag from 'clean-tag';
import { themeGet, space, color, boxShadow, buttonStyle } from 'styled-system';

const getBackground = props =>
  color(props).backgroundColor || buttonStyle(props).backgroundColor;

const Button = styled(tag.button)`
  display: inline-block;
  margin: 0;
  padding: ${themeGet('space.3')} ${themeGet('space.5')};
  font-size: ${themeGet('fontSizes.sm')};
  font-weight: ${themeGet('fontWeights.bold')};
  letter-spacing: ${themeGet('letterSpacings.wide')};
  line-height: ${themeGet('lineHeights.normal')};
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  border: ${themeGet('borders.2')};
  border-color: transparent;
  border-radius: ${themeGet('radii.default')};
  outline: 0;
  transition: ${themeGet('transitions.default')};
  cursor: pointer;
  appearance: none;

  ${buttonStyle}
  ${space}
  ${color}
  ${boxShadow}

  &:hover:not(:disabled) {
    background-color: ${props => darken(0.1, getBackground(props))};
  }

  &:focus {
    box-shadow: ${themeGet('shadows.focus')};
  }

  &:disabled {
    opacity: 0.7;
    cursor: not-allowed;
  }

  ${props => props.rounded && css`
    border-radius: ${themeGet('radii.rounded')};
  `}
`;

Button.propTypes = {
  ...buttonStyle.propTypes,
  ...space.propTypes,
  ...color.propTypes,
  ...boxShadow.propTypes,
  primary: PropTypes.bool,
  rounded: PropTypes.bool,
};

Button.defaultProps = {
  buttonStyle: 'default',
  blacklist: Object.keys(Button.propTypes),
};

export default Button;
