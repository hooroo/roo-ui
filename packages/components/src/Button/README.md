# Button

<!-- STORY -->

## Installation

```shell
$ yarn add @roo-ui/components
```

## Example

```js
import { Button } from '@roo-ui/components';

export default (
  <Button>Hello world</Button>
);
```

## Properties

| Name       | Description                | Type   | Default | Required? |
|:-----------|:---------------------------|:-------|:--------|:----------|
| `primary`  | use the primary color      | `bool` | false   | -         |
| `rounded`  | use a rounded border       | `bool` | false   | -         |
| `block`    | render a full width button | `bool` | false   | -         |
| `disabled` | disable the button         | `bool` | false   | -         |

## Customization

This component can be customized with [styled-system](https://github.com/jxnblk/styled-system) by passing props for [space](https://github.com/jxnblk/styled-system/blob/v2.2.5/README.md#space-responsive), [color](https://github.com/jxnblk/styled-system/blob/v2.2.5/README.md#color-responsive), or [box shadow](https://github.com/jxnblk/styled-system/blob/v2.2.5/README.md#misc).
