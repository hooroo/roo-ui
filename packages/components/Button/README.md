# Button

<!-- STORY -->

## Installation

```shell
$ yarn add @roo-ui/button
```

## Example

```jsx
import Button from '@roo-ui/button';

export default <Button>Hello world</Button>;
```

## Properties

| Name       | Description           | Type   | Default | Required? |
|:-----------|:----------------------|:-------|:--------|:----------|
| `children` | button content        | `node` | -       | ✔︎         |
| `primary`  | use the primary color | `bool` | false   | -         |
| `rounded`  | use a rounded border  | `bool` | false   | -         |
| `outlined` | outlined button       | `bool` | false   | -         |
| `disabled` | disable the button    | `bool` | false   | -         |
