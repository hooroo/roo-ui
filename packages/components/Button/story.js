import React from 'react';
import { storiesOf } from '@storybook/react';
import { withDocs } from 'storybook-readme';
import { text, boolean } from '@storybook/addon-knobs/react';

import Button from './src';
import README from './README.md';

storiesOf('Button', module)
  .addDecorator(withDocs(README))
  .add('default', () => (
    <Button
      primary={boolean('Primary', false)}
      outlined={boolean('Outlined', false)}
      rounded={boolean('Rounded', false)}
      disabled={boolean('Disabled', false)}
    >
      {text('Children', 'Hello world')}
    </Button>
  ));
