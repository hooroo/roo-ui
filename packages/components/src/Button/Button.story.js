import React from 'react';
import { storiesOf } from '@storybook/react';
import { withDocs } from 'storybook-readme';
import { boolean } from '@storybook/addon-knobs/react';

import Button from '.';
import README from './README.md';

storiesOf('Components|Button', module)
  .addDecorator(withDocs(README))
  .add('default', () => (
    <Button
      primary={boolean('Primary', false)}
      rounded={boolean('Rounded', false)}
      block={boolean('Block', false)}
      disabled={boolean('Disabled', false)}
    >
      Hello world
    </Button>
  ));
