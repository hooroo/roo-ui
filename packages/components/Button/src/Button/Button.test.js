import React from 'react';
import { qantas } from '@roo-ui/themes';
import { shallowWithTheme } from '@roo-ui/test-utils';

import Button from '.';

describe('<Button />', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {};
    wrapper = shallowWithTheme(<Button {...props}>Hello world</Button>, qantas);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
