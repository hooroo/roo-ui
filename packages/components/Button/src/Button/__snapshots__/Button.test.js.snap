// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
.c0 {
  display: inline-block;
  margin: 0;
  padding: 0.75rem 1.5rem;
  font-family: QantasCiutadella,sans-serif;
  font-size: 1rem;
  font-weight: 600;
  -webkit-letter-spacing: 0.125em;
  -moz-letter-spacing: 0.125em;
  -ms-letter-spacing: 0.125em;
  letter-spacing: 0.125em;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  border: 2px solid;
  border-color: transparent;
  border-radius: 4px;
  outline: 0;
  -webkit-transition: 200ms ease-in;
  transition: 200ms ease-in;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: white;
  background-color: #323232;
}

.c0:hover:not(:disabled) {
  background-color: #181818;
}

.c0:focus {
  box-shadow: 0 0 0 2px undefined;
}

.c0:disabled {
  opacity: 0.7;
  cursor: not-allowed;
}

<Clean.button
  buttonStyle="default"
  className="c0"
>
  Hello world
</Clean.button>
`;
